quickwit-indexing/src/source/kinesis/shard_consumer.rs

```diff
@@ -0,0 +1,451 @@
+        json!({
+            "stream_name": self.stream_name,
+            "shard_id": self.shard_id,
+            "current_sequence_number": self.state.current_sequence_number,
+            "lag_millis": self.state.lag_millis,
+            "num_bytes_transferred": self.state.num_bytes_transferred,
+            "num_records_transferred": self.state.num_records_transferred,
+        })
+    }
+}
+
+#[async_trait]
+impl AsyncActor for ShardConsumer {
+    async fn initialize(
+        &mut self,
+        ctx: &ActorContext<Self::Message>,
+    ) -> Result<(), ActorExitStatus> {
+        self.state.next_shard_iterator = get_shard_iterator(
+            &*self.kinesis_client,
+            &self.stream_name,
+            &self.shard_id,
+            self.from_sequence_number_exclusive.clone(),
+        )
+        .await?;
+        self.process_message(Loop {}, ctx).await?;
+        Ok(())
+    }
+
+    async fn process_message(
+        &mut self,
+        _message: Loop,
+        ctx: &ActorContext<Self::Message>,
+    ) -> Result<(), ActorExitStatus> {
+        if let Some(shard_iterator) = self.state.next_shard_iterator.take() {
+            let response = get_records(&*self.kinesis_client, shard_iterator).await?;
+            self.state.lag_millis = response.millis_behind_latest.clone();
+            self.state.next_shard_iterator = response.next_shard_iterator;
+
+            if !response.records.is_empty() {
+                self.state.current_sequence_number = response
+                    .records
+                    .last()
+                    .map(|record| record.sequence_number.clone());
+                self.state.num_bytes_transferred += response
+                    .records
+                    .iter()
+                    .map(|record| record.data.len() as u64)
+                    .sum::<u64>();
+                self.state.num_records_transferred += response.records.len() as u64;
+
+                let message = ShardConsumerMessage::Records {
+                    shard_id: self.shard_id.clone(),
+                    records: response.records,
+                    lag_millis: response.millis_behind_latest,
+                };
+                ctx.send_message(&self.sink, message).await?;
+            }
+            if let Some(children) = response.child_shards {
+                let shard_ids: Vec<String> = children
+                    .into_iter()
+                    // Filter out duplicate message when two shards are merged.
+                    .filter(|child| child.parent_shards.first() == Some(&self.shard_id))
+                    .map(|child| child.shard_id)
+                    .collect();
+                if !shard_ids.is_empty() {
+                    let message = ShardConsumerMessage::ChildShards(shard_ids);
+                    ctx.send_message(&self.sink, message).await?;
+                }
+            }
+            if self.eof_enabled && response.millis_behind_latest.unwrap_or(-1) == 0 {
```

Comment:
I'd rather have `== Some(0)` instead of the unwrap_or(-1)

Reply:
Ok.